# TASCA MÈTODES DE MOSTREIG

## NIVELL1

### *Exercici 1*

Agafa un conjunt de dades de tema esportiu que t'agradi. Realitza un mostreig de les dades generant una mostra aleatòria simple i una mostra sistemàtica.

+ *Jo he triat un data set de tenis amb aquesta URL https://www.kaggle.com/tbrownlow3/atp-tennis?select=Match.csv*

  *En aquest data set es recullen dades sobre els 4 tornejos (Gran Slam) més importants del calendari anual des de l'any 2000
   fins a 2019 inclòs*
   
   *La variable sobre la que vull treballar és la durada dels partits, que està recollida per les diferents observacions a la 
    columna amb índex 8 (" match_minutes")
   
   

In [1]:
%matplotlib inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import matlab as mat
import random

In [2]:
from matlab import random
from random import randint

In [3]:
tenis_df = pd.read_csv('Match.csv')

In [4]:
tenis_df

match_id          round        date  avg_minutes_game  \
0        m_2019_A_0      1st Round  2019-01-14              4.81   
1        m_2019_A_1      1st Round  2019-01-14              3.95   
2        m_2019_A_2      1st Round  2019-01-14              4.86   
3        m_2019_A_3      1st Round  2019-01-14              4.43   
4        m_2019_A_4      1st Round  2019-01-14              4.37   
...             ...            ...         ...               ...   
10155  m_2000_W_122  Quarterfinals  2000-06-26              3.95   
10156  m_2000_W_123  Quarterfinals  2000-06-26              4.26   
10157  m_2000_W_124     Semifinals  2000-06-26              3.74   
10158  m_2000_W_125     Semifinals  2000-06-26              3.30   
10159  m_2000_W_126      The Final  2000-06-26              4.24   

       avg_seconds_point  avg_minutes_set       tournament  year  \
0                   46.8             43.3  Australian Open  2019   
1                   35.7             40.5  Australian Open  2019   
2                   42.3             43.8  Australian Open  2019   
3                   43.3             41.0  Australian Open  2019   
4                   39.7             43.7  Australian Open  2019   
...                  ...              ...              ...   ...   
10155               37.3             41.5        Wimbledon  2000   
10156               39.0             48.3        Wimbledon  2000   
10157               37.2             39.6        Wimbledon  2000   
10158               34.7             33.0        Wimbledon  2000   
10159               35.7             44.5        Wimbledon  2000   

       match_minutes  
0              173.0  
1              162.0  
2              175.0  
3              164.0  
4              131.0  
...              ...  
10155          166.0  
10156          145.0  
10157          198.0  
10158           99.0  
10159          178.0  

[10160 rows x 9 columns]

In [5]:
tenis_df.describe()

avg_minutes_game  avg_seconds_point  avg_minutes_set          year  \
count      10114.000000       10115.000000     10078.000000  10160.000000   
mean           4.058700          38.248384        39.223477   2009.500000   
std            0.890068           8.493856        10.129553      5.766565   
min            0.000000           0.000000         0.000000   2000.000000   
25%            3.730000          36.000000        34.400000   2004.750000   
50%            4.110000          38.800000        39.300000   2009.500000   
75%            4.500000          41.700000        44.700000   2014.250000   
max           21.500000         390.000000       133.000000   2019.000000   

       match_minutes  
count   10117.000000  
mean      142.343284  
std        52.397425  
min         0.000000  
25%       108.000000  
50%       138.000000  
75%       176.000000  
max       665.000000

+ *Genero una mostra aleatòria simple amb n observacions*

In [6]:
simple_tenis_df = tenis_df.sample(n = 500)
simple_tenis_df 

match_id          round        date  avg_minutes_game  \
6397   m_2009_U_47      1st Round  2009-02-09              4.15   
4849   m_2001_F_23      1st Round  2001-05-28              4.12   
358   m_2017_A_104      3rd Round  2017-01-21              3.22   
6409   m_2009_U_59      1st Round  2009-02-09              3.64   
3890   m_2009_F_80      2nd Round  2009-05-28              5.93   
...            ...            ...         ...               ...   
1198   m_2010_A_55      1st Round  2010-01-19              5.23   
2406  m_2001_A_120  Quarterfinals  2001-01-15              3.81   
1665   m_2006_A_14      1st Round  2006-01-16              4.96   
7243    m_2002_U_4      1st Round  2002-08-26              4.60   
7184   m_2003_U_72      2nd Round  2003-08-28              4.90   

      avg_seconds_point  avg_minutes_set       tournament  year  match_minutes  
6397               40.3             45.7          US Open  2009          137.0  
4849               38.9             34.3      French Open  2001          103.0  
358                36.3             29.0  Australian Open  2017           87.0  
6409               35.6             32.8          US Open  2009          131.0  
3890               45.4             59.3      French Open  2009          178.0  
...                 ...              ...              ...   ...            ...  
1198               47.9             58.6  Australian Open  2010          293.0  
2406               38.3             39.3  Australian Open  2001          118.0  
1665               44.6             62.0  Australian Open  2006          124.0  
7243               40.4             49.5          US Open  2002          198.0  
7184               41.7             47.8          US Open  2003          191.0  

[500 rows x 9 columns]

In [7]:
simple_tenis_df.describe()

avg_minutes_game  avg_seconds_point  avg_minutes_set         year  \
count        499.000000         499.000000       498.000000   500.000000   
mean           4.076373          38.357715        39.324699  2009.376000   
std            0.782079           6.519895         9.392562     5.907658   
min            0.000000           0.000000         0.000000  2000.000000   
25%            3.720000          36.200000        34.300000  2004.000000   
50%            4.100000          38.700000        39.000000  2009.000000   
75%            4.500000          41.400000        44.700000  2015.000000   
max            8.080000          76.100000        78.600000  2019.000000   

       match_minutes  
count     499.000000  
mean      140.517034  
std        49.994189  
min         0.000000  
25%       109.000000  
50%       135.000000  
75%       168.500000  
max       393.000000

+ *Ara genero una mostra aleatòria sistemàtica.
   Primer obtinc un un número random, que serà el punt de partida des del qual jo fixo arbitràriament el "step" o salt,
   amb el que  s'aniran seleccionant les observacions (files)

In [8]:
random_number = np.random.randint(1, 50)
random_number

43

In [9]:
step = 97
systematic_tenis_df = tenis_df[random_number : len(tenis_df) : step]
systematic_tenis_df

match_id      round        date  avg_minutes_game  \
43      m_2019_A_43  1st Round  2019-01-15              3.80   
140     m_2018_A_13  1st Round  2018-01-15              4.08   
237    m_2018_A_110  3rd Round  2018-01-20              5.42   
334     m_2017_A_80  2nd Round  2017-01-19              5.00   
431     m_2016_A_50  1st Round  2016-01-19              4.03   
...             ...        ...         ...               ...   
9743    m_2003_W_91  2nd Round  2003-06-26              3.70   
9840    m_2002_W_61  1st Round  2002-06-24              3.64   
9937    m_2001_W_31  1st Round  2001-06-25              3.74   
10034    m_2000_W_1  1st Round  2000-06-26              2.83   
10131   m_2000_W_98  3rd Round  2000-06-26              3.17   

       avg_seconds_point  avg_minutes_set       tournament  year  \
43                  37.5             33.3  Australian Open  2019   
140                 38.7             32.7  Australian Open  2018   
237                 46.2             47.0  Australian Open  2018   
334                 44.1             35.0  Australian Open  2017   
431                 38.5             41.7  Australian Open  2016   
...                  ...              ...              ...   ...   
9743                36.4             37.0        Wimbledon  2003   
9840                29.1             34.0        Wimbledon  2002   
9937                36.4             39.3        Wimbledon  2001   
10034               29.6             22.7        Wimbledon  2000   
10131               31.4             25.3        Wimbledon  2000   

       match_minutes  
43             133.0  
140             98.0  
237            141.0  
334            105.0  
431            125.0  
...              ...  
9743           111.0  
9840           102.0  
9937           157.0  
10034           68.0  
10131           76.0  

[105 rows x 9 columns]

In [10]:
systematic_tenis_df.describe()

avg_minutes_game  avg_seconds_point  avg_minutes_set         year  \
count        105.000000         105.000000       105.000000   105.000000   
mean           4.056762          38.392381        38.987619  2009.533333   
std            0.800283           6.862547         9.353359     5.824462   
min            0.000000           0.000000         0.000000  2000.000000   
25%            3.770000          36.400000        34.700000  2005.000000   
50%            4.080000          38.900000        38.300000  2010.000000   
75%            4.480000          41.700000        44.800000  2015.000000   
max            5.630000          50.000000        64.800000  2019.000000   

       match_minutes  
count     105.000000  
mean      135.923810  
std        50.588212  
min         0.000000  
25%       107.000000  
50%       129.000000  
75%       165.000000  
max       259.000000

## NIVELL 2

### *Exercici 2

Continua amb el conjunt de dades de tema esportiu i genera una mostra estratificada i una mostra utilitzant SMOTE (Synthetic Minority Oversampling Technique).

+ *En primer lloc miro com són de diferents les observacions per cada torneig dels recollits pel data set*

In [11]:
tenis_df['tournament'].value_counts()

US Open            2540
Wimbledon          2540
Australian Open    2540
French Open        2540
Name: tournament, dtype: int64

+ *Els 4 tornejos tenen idèntiques observacions.Per tant no sembla apropiat aplicar un mostreig estratificat*

   *Miro la columna "round" on és més evident que trobaré diferent número d'observacions segons sigui 1a, 2a,..., semifinals
    o finals*

In [12]:
tenis_df['round'].unique()

array(['1st Round', '2nd Round', '3rd Round', '4th Round',
       'Quarterfinals', 'Semifinals', 'The Final'], dtype=object)

In [13]:
tenis_df['round'].value_counts()

1st Round        5120
2nd Round        2560
3rd Round        1280
4th Round         640
Quarterfinals     320
Semifinals        160
The Final          80
Name: round, dtype: int64

+ *I genero la mostra aleatòria per estrats ("rounds")

In [14]:
from sklearn.model_selection import train_test_split
Xtrain, Xtest = train_test_split(tenis_df, test_size=0.20, stratify=tenis_df[['round']])
Xtrain

match_id      round        date  avg_minutes_game  avg_seconds_point  \
7804  m_2018_W_57  1st Round  2018-03-07              3.48               33.7   
970   m_2012_A_81  2nd Round  2012-01-19              4.97               46.7   
4716  m_2002_F_17  1st Round  2002-05-27              5.39               40.1   
7044  m_2004_U_59  1st Round  2004-01-09              4.45               39.7   
2717  m_2018_F_50  1st Round  2018-05-29              4.30               43.8   
...           ...        ...         ...               ...                ...   
1614  m_2007_A_90  2nd Round  2007-01-18              3.87               41.9   
9240  m_2007_W_96  3rd Round  2007-06-29              4.67               40.4   
9164  m_2007_W_20  1st Round  2007-06-26              3.83               34.2   
4733  m_2002_F_34  1st Round  2002-05-27              4.34               39.1   
4779  m_2002_F_80  2nd Round  2002-05-27              4.82               42.1   

      avg_minutes_set       tournament  year  match_minutes  
7804             33.7        Wimbledon  2018          101.0  
970              48.0  Australian Open  2012          144.0  
4716             50.3      French Open  2002          151.0  
7044             39.2          US Open  2004          196.0  
2717             38.7      French Open  2018          116.0  
...               ...              ...   ...            ...  
1614             38.7  Australian Open  2007          116.0  
9240             45.5        Wimbledon  2007          182.0  
9164             34.5        Wimbledon  2007          138.0  
4733             44.5      French Open  2002          178.0  
4779             45.8      French Open  2002          183.0  

[8128 rows x 9 columns]

In [15]:
Xtrain['round'].unique()

array(['1st Round', '2nd Round', '3rd Round', 'Quarterfinals',
       '4th Round', 'The Final', 'Semifinals'], dtype=object)

In [16]:
Xtrain['round'].describe()

count          8128
unique            7
top       1st Round
freq           4096
Name: round, dtype: object

+ *Per fer SMOTE haig d'instalar "imblearn" i imprortar SMOTE amb "from imblearn.oversampling" al notebook*

In [17]:
import imblearn

In [18]:
from imblearn.over_sampling import SMOTE

+ *Genero un sub data frame amb les files on el valor de la columna "round" sigui "Semifinals" o "The Final"*

In [19]:
tenis_df2 = tenis_df[(tenis_df['round'] == 'Semifinals') | (tenis_df['round'] == 'The Final')]

In [20]:
tenis_df2

match_id       round        date  avg_minutes_game  \
124    m_2019_A_124  Semifinals  2019-01-24              4.42   
125    m_2019_A_125  Semifinals  2019-01-25              3.77   
126    m_2019_A_126   The Final  2019-01-27              4.77   
251    m_2018_A_124  Semifinals  2018-01-25              4.93   
252    m_2018_A_125  Semifinals  2018-01-26              4.43   
...             ...         ...         ...               ...   
10031  m_2001_W_125  Semifinals  2001-06-25              3.75   
10032  m_2001_W_126   The Final  2001-06-25              3.57   
10157  m_2000_W_124  Semifinals  2000-06-26              3.74   
10158  m_2000_W_125  Semifinals  2000-06-26              3.30   
10159  m_2000_W_126   The Final  2000-06-26              4.24   

       avg_seconds_point  avg_minutes_set       tournament  year  \
124                 50.5             35.3  Australian Open  2019   
125                 39.2             27.7  Australian Open  2019   
126                 52.4             41.3  Australian Open  2019   
251                 43.6             46.0  Australian Open  2018   
252                 41.3             62.0  Australian Open  2018   
...                  ...              ...              ...   ...   
10031               36.9             36.0        Wimbledon  2001   
10032               36.0             36.4        Wimbledon  2001   
10157               37.2             39.6        Wimbledon  2000   
10158               34.7             33.0        Wimbledon  2000   
10159               35.7             44.5        Wimbledon  2000   

       match_minutes  
124            106.0  
125             83.0  
126            124.0  
251            138.0  
252             62.0  
...              ...  
10031          180.0  
10032          182.0  
10157          198.0  
10158           99.0  
10159          178.0  

[240 rows x 9 columns]

+ *I comprovo que ha funcionat*


In [23]:
tenis_df2['round'].unique()

array(['Semifinals', 'The Final'], dtype=object)

+ *Em desfaig de les columnes amb valors "string" (no numériques)*

In [24]:
tenis_df2 = tenis_df2.drop(columns =['match_id', 'date', 'tournament'])

In [25]:
tenis_df2

round  avg_minutes_game  avg_seconds_point  avg_minutes_set  year  \
124    Semifinals              4.42               50.5             35.3  2019   
125    Semifinals              3.77               39.2             27.7  2019   
126     The Final              4.77               52.4             41.3  2019   
251    Semifinals              4.93               43.6             46.0  2018   
252    Semifinals              4.43               41.3             62.0  2018   
...           ...               ...                ...              ...   ...   
10031  Semifinals              3.75               36.9             36.0  2001   
10032   The Final              3.57               36.0             36.4  2001   
10157  Semifinals              3.74               37.2             39.6  2000   
10158  Semifinals              3.30               34.7             33.0  2000   
10159   The Final              4.24               35.7             44.5  2000   

       match_minutes  
124            106.0  
125             83.0  
126            124.0  
251            138.0  
252             62.0  
...              ...  
10031          180.0  
10032          182.0  
10157          198.0  
10158           99.0  
10159          178.0  

[240 rows x 6 columns]

+ *I converteixo els valors de la columna Semofinals = 0 i The Final = 1*

In [26]:
tenis_df2['round'] = [0 if x == 'Semifinals' else 1 for x in tenis_df2['round']]

In [27]:
tenis_df2

round  avg_minutes_game  avg_seconds_point  avg_minutes_set  year  \
124        0              4.42               50.5             35.3  2019   
125        0              3.77               39.2             27.7  2019   
126        1              4.77               52.4             41.3  2019   
251        0              4.93               43.6             46.0  2018   
252        0              4.43               41.3             62.0  2018   
...      ...               ...                ...              ...   ...   
10031      0              3.75               36.9             36.0  2001   
10032      1              3.57               36.0             36.4  2001   
10157      0              3.74               37.2             39.6  2000   
10158      0              3.30               34.7             33.0  2000   
10159      1              4.24               35.7             44.5  2000   

       match_minutes  
124            106.0  
125             83.0  
126            124.0  
251            138.0  
252             62.0  
...              ...  
10031          180.0  
10032          182.0  
10157          198.0  
10158           99.0  
10159          178.0  

[240 rows x 6 columns]

In [29]:
tenis_df2['round'].value_counts()

0    160
1     80
Name: round, dtype: int64

+ *I faig una ullada a la columna amb ".describe()"*

In [30]:
tenis_df2['round'].describe()

count    240.000000
mean       0.333333
std        0.472390
min        0.000000
25%        0.000000
50%        0.000000
75%        1.000000
max        1.000000
Name: round, dtype: float64

+ *I  una ullada al data frame amb ".describe()"*

In [31]:
tenis_df2.describe()

round  avg_minutes_game  avg_seconds_point  avg_minutes_set  \
count  240.000000        240.000000          240.00000       240.000000   
mean     0.333333          4.458375           42.04375        43.796667   
std      0.472390          1.005532            8.64380        11.177008   
min      0.000000          0.000000            0.00000         0.000000   
25%      0.000000          4.030000           38.70000        38.000000   
50%      0.000000          4.490000           42.50000        44.300000   
75%      1.000000          5.065000           46.72500        49.925000   
max      1.000000          6.760000           57.40000        79.200000   

              year  match_minutes  
count   240.000000     240.000000  
mean   2009.500000     164.675000  
std       5.778332      60.417746  
min    2000.000000       0.000000  
25%    2004.750000     126.500000  
50%    2009.500000     161.500000  
75%    2014.250000     201.750000  
max    2019.000000     396.000000

+ *I ara sí, amb "SMOTE" genero un mostreig on les observacions amb el valor "The Final" s'igualin amb les de "Semifinals"*

   *Han de passar de 80 observacions a 160*

In [32]:
smote = SMOTE(sampling_strategy = 'minority')
X_sm, y_sm = smote.fit_resample(tenis_df2, tenis_df2['round'])

In [33]:
X_sm

round  avg_minutes_game  avg_seconds_point  avg_minutes_set  year  \
0        0          4.420000          50.500000        35.300000  2019   
1        0          3.770000          39.200000        27.700000  2019   
2        1          4.770000          52.400000        41.300000  2019   
3        0          4.930000          43.600000        46.000000  2018   
4        0          4.430000          41.300000        62.000000  2018   
..     ...               ...                ...              ...   ...   
315      1          3.677244          33.554951        36.075044  2001   
316      1          4.334946          37.984643        44.500000  2001   
317      1          5.059753          47.492578        45.561848  2005   
318      1          4.697060          47.429064        45.515106  2014   
319      1          5.099457          46.113343        47.055059  2009   

     match_minutes  
0       106.000000  
1        83.000000  
2       124.000000  
3       138.000000  
4        62.000000  
..             ...  
315     144.100174  
316     178.000000  
317     182.247393  
318     136.645318  
319     188.167888  

[320 rows x 6 columns]

In [34]:
X_sm['round'].value_counts()

0    160
1    160
Name: round, dtype: int64

## NIVELL 3

### *Exercici 3*

Continua amb el conjunt de dades de tema esportiu i genera una mostra utilitzant el mètode Reservoir sampling.

In [35]:
tenis_df

match_id          round        date  avg_minutes_game  \
0        m_2019_A_0      1st Round  2019-01-14              4.81   
1        m_2019_A_1      1st Round  2019-01-14              3.95   
2        m_2019_A_2      1st Round  2019-01-14              4.86   
3        m_2019_A_3      1st Round  2019-01-14              4.43   
4        m_2019_A_4      1st Round  2019-01-14              4.37   
...             ...            ...         ...               ...   
10155  m_2000_W_122  Quarterfinals  2000-06-26              3.95   
10156  m_2000_W_123  Quarterfinals  2000-06-26              4.26   
10157  m_2000_W_124     Semifinals  2000-06-26              3.74   
10158  m_2000_W_125     Semifinals  2000-06-26              3.30   
10159  m_2000_W_126      The Final  2000-06-26              4.24   

       avg_seconds_point  avg_minutes_set       tournament  year  \
0                   46.8             43.3  Australian Open  2019   
1                   35.7             40.5  Australian Open  2019   
2                   42.3             43.8  Australian Open  2019   
3                   43.3             41.0  Australian Open  2019   
4                   39.7             43.7  Australian Open  2019   
...                  ...              ...              ...   ...   
10155               37.3             41.5        Wimbledon  2000   
10156               39.0             48.3        Wimbledon  2000   
10157               37.2             39.6        Wimbledon  2000   
10158               34.7             33.0        Wimbledon  2000   
10159               35.7             44.5        Wimbledon  2000   

       match_minutes  
0              173.0  
1              162.0  
2              175.0  
3              164.0  
4              131.0  
...              ...  
10155          166.0  
10156          145.0  
10157          198.0  
10158           99.0  
10159          178.0  

[10160 rows x 9 columns]

+ *Utilitzo l'algorisme que hi ha al recurs "The 5 Sampling Algorithms every Data Scientist need to know"*

In [39]:
import random
def generator(max):
    number = 1
    while number < max:
        number += 1
        yield number

stream = tenis_df['match_minutes']

k=100
reservoir = []
for i, element in enumerate(tenis_df['match_minutes']):
    if i+1<= k:
        reservoir.append(element)
    else:
        probability = k/(i+1)
        if random.random() < probability:
           
             reservoir[random.choice(range(0,k))] = element

In [43]:
reservoir

[175.0,
 113.0,
 109.0,
 177.0,
 87.0,
 121.0,
 290.0,
 185.0,
 169.0,
 114.0,
 21.0,
 184.0,
 229.0,
 251.0,
 154.0,
 128.0,
 144.0,
 148.0,
 137.0,
 121.0,
 121.0,
 106.0,
 211.0,
 125.0,
 140.0,
 208.0,
 168.0,
 131.0,
 128.0,
 82.0,
 0.0,
 127.0,
 118.0,
 203.0,
 98.0,
 123.0,
 257.0,
 241.0,
 132.0,
 98.0,
 118.0,
 101.0,
 124.0,
 168.0,
 116.0,
 120.0,
 116.0,
 105.0,
 128.0,
 101.0,
 127.0,
 119.0,
 97.0,
 130.0,
 98.0,
 146.0,
 164.0,
 136.0,
 195.0,
 101.0,
 108.0,
 116.0,
 116.0,
 105.0,
 84.0,
 165.0,
 193.0,
 91.0,
 141.0,
 172.0,
 144.0,
 123.0,
 121.0,
 89.0,
 162.0,
 101.0,
 150.0,
 121.0,
 198.0,
 0.0,
 112.0,
 156.0,
 200.0,
 147.0,
 172.0,
 107.0,
 107.0,
 141.0,
 143.0,
 101.0,
 152.0,
 132.0,
 118.0,
 184.0,
 217.0,
 173.0,
 101.0,
 106.0,
 115.0,
 104.0]

+ *I comprovo que tinc els 100 valors esperats quan hem fixat k = 100 i l'estadística bàsica  de la variable*

In [46]:
len(reservoir)

100

In [63]:
a, b, c = min(reservoir), max(reservoir), sum(reservoir)/len(reservoir)
print('El valor mínim es {} el valor màxim és {} i la duració mitjana dels partits és de {}'.format(a, b, c))

El valor mínim es 0.0 el valor màxim és 290.0 i la duració mitjana dels partits és de 136.72
